NeoMathEngine/src/CPU/CpuMathEngineDnnDropout.cpp

```diff
@@ -41,28 +57,126 @@ void CCpuMathEngine::Dropout( const CDropoutDesc& dropoutDesc, const CFloatHandl
 	const int batchWidth = input.ObjectCount() / batchLength;
 	const int maskSize = batchWidth * objectSize;
 
-	ASSERT_EXPR( desc.Mask.Size() == maskSize );
+	CCpuRandom random(desc.Seed);
+	CIntArray<CSeedDropoutDesc::MaskAlign> generated;
 
-	if( !desc.IsSpatial ) {
-		MultiplyMatrixByDiagMatrix( inputData, batchLength, maskSize, desc.Mask.GetHandle(),
-			outputData, output.BlobSize() );
-		return;
-	}
+	const int inputObjectSize = input.ObjectSize();
+	const float* inputPointer = GetRaw(inputData);
+	float* outputPointer = GetRaw(outputData);
+	float* mask = GetRaw(desc.Mask->GetHandle());
+	const unsigned int threshold = desc.Threshold;
+	const float value = desc.Value;
+	constexpr int cacheSize = CSeedDropoutDesc::CacheSize;
+	constexpr int maskAlign = CSeedDropoutDesc::MaskAlign;
+	constexpr int numOfGenerations = CSeedDropoutDesc::NumOfGenerations;
 
-	CFloatHandle currInput = inputData;
-	CFloatHandle currOutput = outputData;
-	for( int i = 0; i < input.ObjectCount(); ++i ) {
-		MultiplyMatrixByDiagMatrix( currInput, input.ObjectSize() / objectSize, objectSize,
-			desc.Mask.GetHandle() + ( i % batchWidth ) * objectSize, currOutput, input.ObjectSize() );
-		currInput += input.ObjectSize();
-		currOutput += input.ObjectSize();
-	}
-}
+	if(!desc.IsSpatial) {
+		const int numOfIter = (maskSize + cacheSize - 1) / cacheSize;
+		const int currSize = maskSize - (numOfIter - 1) * cacheSize;
+		const int lastGenerations = (currSize + (maskAlign - 1)) / maskAlign;
 
-CDropoutDesc* CCpuMathEngine::InitDropout( float rate, bool isSpatial, bool isBatchwise,
-	const CBlobDesc& input, const CBlobDesc& output, int seed )
-{
-	return new CMathEngineDropoutDesc( mathEngine(), rate, isSpatial, isBatchwise, input, output, seed );
+		for(int i = 0; i < numOfIter - 1; ++i) {
+			const float* first = inputPointer;
+			float* result = outputPointer;
+
+			int idx = 0;
```

Comment:
I have seen this code four times here! Let's make a function.

Reply:
Also it was my ask. When we excepted nested functions calls, the performance become better